File: Samu/Pr04/src/Usuario.java
import java.util.*;

public class Usuario {
    private static String BD_SERVER = "localhost";
    private static String BD_NAME = "GI1819";

    private String nombre;
    private String password;
    private Rol rol;

    public static List<Usuario> ListaUsuarios() {
        // Retorna una lista con todos los objetos de la clase almacenados en la base de datos
        List<Usuario> lista = new ArrayList<Usuario>();
        BD miBD = new BD(BD_SERVER, BD_NAME);

        for (Object[] tupla : miBD.Select("SELECT * FROM tUsuario;")) {
            Usuario u = new Usuario((String) tupla[0], (String) tupla[1]);
            lista.add(u);
        }
        return lista;
    }

    public Usuario(String n, String p) {
        // Crea el objeto cargando sus valores de la base de datos
        // Si la password almacenada no se corresponde con la suministrada se elevara una excepcion
        BD miBD = new BD(BD_SERVER, BD_NAME);
        Object[] tupla = miBD.Select(String.format("Select * from tUsuario where nombre='%s';",
                n)).get(0);
        if(tupla[1].equals(p)){
            this.nombre = (String) tupla[0];
            this.password = (String) tupla[1];
            this.rol = new Rol((String) tupla[2]);
        } else {
            throw new Error("Usuario y contraseña no coinciden");
        }
    }

    public Usuario(String n, String p, Rol r) {
        // Crea el objeto y lo inserta en la base de datos
        BD miBD = new BD(BD_SERVER, BD_NAME);
        miBD.Insert(String.format("Insert into tUsuario(nombre, password, rolName) values ('%s', '%s', '%s');",
                n, p, r.getRolName()));
        this.nombre = n;
        this.password = p;
        this.rol = r;
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String value) {

        // Actualiza el atributo en memoria y en la base de datos
        BD miBD = new BD(BD_SERVER, BD_NAME);
        miBD.Update(String.format("update tUsuario set nombre = '%s' where nombre = '%s';", value, this.nombre));
        this.nombre = value;

    }

    public void borraUsuario() {
        // Actualiza el atributo en memoria y en la base de datos
        BD miBD = new BD(BD_SERVER, BD_NAME);
        miBD.Delete(String.format("delete from tUsuario where nombre = '%s';", this.nombre));
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String value) {
        // Actualiza el atributo en memoria y en la base de datos
        BD miBD = new BD(BD_SERVER, BD_NAME);
        miBD.Update(String.format("update tUsuario set password = '%s' where nombre = '%s';", value, this.nombre));
        this.password = value;

    }

    public Rol getRol() {
        return rol;
    }

    public void setRol() {
        throw new Error("Error: Un usuario no puede cambiar el rol de si mismo.");
    }

    public void ModiRol(Usuario u, Rol r) {
        // Si el objeto actual es administrador
        if(this.rol.getAdmin()){
            BD miBD = new BD(BD_SERVER, BD_NAME);
            miBD.Update(String.format("update tUsuario set rolName = '%s' where nombre = '%s';", r.getRolName(),
                    u.nombre));
            u.rol = r;
        }

    }

    public boolean AccesoPantalla(String p) {
        return rol.Acceso(p);
    }

    public boolean ModificaPantalla(String p) {
        return rol.Modificacion(p);
    }

    public String toString() {

        return nombre + "\t" + password + "\t" + rol.getRolName();
    }
}

File: Aleks/Practica6/Usuario.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using BDLibrary;

namespace HIS
{
    public class Usuario
    {
        private static string BD_SERVER = Properties.Settings.Default.BD_SERVER;
        private static string BD_NAME = Properties.Settings.Default.BD_NAME;

        private string nombre;
        private string password;
        private Rol rol;


	    public static List<Usuario> ListaUsuarios()
	    {
		    List<Usuario> lista = new List<Usuario>();

            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);

            foreach (object[] tupla in miBD.Select("SELECT nombre, password FROM tUsuario;"))
            {
                string n = (string)tupla[0];
                string p = (string)tupla[1];
                lista.Add(new Usuario(n,p));
            }

		    return lista;
	    }

        public Usuario(string n, string p)
        {

            try
            {
                SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
                object[] tupla = miBD.Select("SELECT * FROM tUsuario WHERE nombre = '" + n + "';")[0];

                nombre = (string)tupla[0];
                password = (string)tupla[1];
                rol = new Rol((string)tupla[2]);

                if (!password.Equals(p))
                {
                    nombre = password = null;
                    rol = null;
                    throw new Error("Usuario o Contraseña Incorrecta");
                }
            }
            catch (Exception ex)
            {
                throw new Error("Usuario o Contraseña Incorrecta: " + ex.Message);
            }
                
        }

        public Usuario(string n, string p, Rol r)
        {

            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);

            miBD.Insert("INSERT INTO tUsuario VALUES('" + n + "', '"
                + p + "', '" + r.RolName + "');");

            nombre = n;
            password = p;
            rol = r;

        }

        public string Nombre
        {
            get { return nombre; }
            set {
                    SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
                    miBD.Update("UPDATE tUsuario SET nombre = '" + value
                            + "' WHERE nombre = '" + nombre + "';");
                    nombre = value;
                }
        }

        public string Password
        {
            get { return password; }
            set {
                    SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
                    miBD.Update("UPDATE tUsuario SET password = '" + value
                            + "' WHERE nombre = '" + nombre + "';");
                    password = value;
                }
        }

        public Rol MyRol
        {
            get { return rol; }
        }

        public void ModiRol(Usuario u, Rol r)
        {
            if (!this.rol.Admin) 
                throw new Error("El usuario " + this.Nombre 
                    + " no puede cambiar el rol del usuario " + u.Nombre);
            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
            miBD.Update("UPDATE tUsuario SET rolName = '" + r.RolName
                    + "' WHERE nombre = '" + u.Nombre + "';");
            u.rol = r;

        }

        public void BorrarUsuario()
        {
            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
            miBD.Delete("DELETE tUsuario WHERE nombre = '" + nombre + "';");
            nombre = password = null;
            rol = null;
        }


        public bool AccesoPantalla(string p)
        {
            return rol.Acceso(p);
        }

        public bool ModificaPantalla(string p)
        {
            return rol.Modificacion(p);
        }

    }
}
